signer/src/transaction_signer.rs

```diff
@@ -146,6 +147,9 @@ pub struct TxSignerEventLoop<Context, Network, Rng> {
     /// during DKG using the FROST algorithm. This is then used during the
     /// verification of the Stacks rotate-keys transaction.
     pub dkg_verification_state_machines: LruCache<StateMachineId, dkg::verification::StateMachine>,
+    /// Stacks transactions signed during a bitcoin tenure. We don't allow
+    /// signing for the same request multiple times in a tenure.
+    pub stacks_sign_request: LruCache<model::BitcoinBlockHash, HashSet<StacksSignRequestId>>,
```

Comment:
I guess I'm wondering if we even need an `LruCache` here at all, and instead just have a `HashSet<StacksSignRequestId>` that gets `clear()`'d on `BitcoinBlockObserved` event. Or something along those lines anyway, as this is only ever modified in a code path where the `msg.chain_tip` == the signer's canonical tip (which will change every time a `BitcoinBlockObserved` arrives).

This is fine too (functionally), my above thought is more implicit which may be less desirable given some of the complexity in this file, just felt a little overkill at first glance and would save you the added `&mut self` below 🤷

Reply:
Yeah I think we should be fine with a cache of max size 1, so then we can just drop the LRU. But I'm not sure if we can rely on `BitcoinBlockObserved`, as it's not always emitted (if we err and `continue` in the block observer) and we emit it "later" than when we processed the bitcoin blocks, so we may risk getting a message that passes validation on the tx signer before we got that tenure `BitcoinBlockObserved`, which would mess things up. I guess we could just keep what's the tenure we are currently using for the hashset, but then are we really gaining something wrt using the LRU with size of 1?

Reply:
The above was mostly a thought as LRU's are pretty heavyweight in comparison, but I'm fine leaving it as an LRU(2); LRU(1) feels a _bit_ riskier in-case somehow some message combination makes it through and manages to evict the "correct" hashset.